core/src/main/scala/org/http4s/util/encoding/UriCodingUtils.scala

```diff
@@ -0,0 +1,271 @@
+    // We are decoding, but the encoding procedure is as follows:
+    // replace ' ' with '+'
+    // do not encode * - . 0-9 A-Z _ a-z
+    // do percent-encode everything else
+
+    def decode(s: String) = percentDecode(PctEncoded(s.replace("+", " ")))
+
+    if (encoded.isEmpty)
+      Vector("" -> None)
+    else
+      encoded.split("&", Integer.MAX_VALUE).toVector.map {
+        string =>
+          val (name, value) =
+            string.split("=", 2) match {
+              case Array(name)        => (name, None)
+              case Array(name, value) => (name, Some(value))
+            }
+          (decode(name), value.map(decode))
+      }
+
+  }
+
+  def split[C[_]<:SeqLike[_, C[A]],A](payload: C[A], delimiter: A): Vector[C[A]] = {
+    @tailrec def loop(mark: Int, pos: Int, acc: Vector[C[A]]): Vector[C[A]] = {
+      if (pos < payload.size)
+        if (payload(pos) == delimiter)
+          loop(pos+1, pos+1, acc :+ payload.slice(mark, pos))
+        else
+          loop(mark, pos+1, acc)
+      else acc :+ payload.slice(mark, pos)
+    }
+    loop(0, 0, Vector())
+  }
+
+  def split[@specialized(Byte, Char, Int) A](payload: Array[A], delimiter: A): Vector[Array[A]] = {
+    @tailrec def loop(mark: Int, pos: Int, acc: Vector[Array[A]]): Vector[Array[A]] = {
+      if (pos < payload.length)
+        if (payload(pos) == delimiter)
+          loop(pos+1, pos+1, acc :+ payload.slice(mark, pos))
+        else
+          loop(mark, pos+1, acc)
+      else acc :+ payload.slice(mark, pos)
+    }
+    loop(0, 0, Vector())
+  }
+
+//  // doesn't support "isindex"
+//  // does support charset!
+//  def w3cHtml5FormUrlDecode(payload: Array[Byte], defaultEncoding: Charset = Utf8): Vector[(String, Option[String])] = {
+//    def decode(s: String) = percentDecode(PercentEncodedString(s.replace("+", " ")))
+//
+//    if (payload.isEmpty) Vector()
+//    else {
+//      val kvPairs: Vector[(Array[Byte], Option[Array[Byte]])] =
+//        split(payload, '&'.toByte).map {
+//          kv =>
+//            val delimIndex = kv.indexWhere(_ == '='.toByte)
+//            if (delimIndex < 0) kv -> None
+//            else kv.slice(0, delimIndex) -> Some(kv.slice(delimIndex+1, kv.length))
+//        }
+//      val charset =
+//        kvPairs
+//          .find(kv => kv._1 == "_charset_".getBytes(Utf8))
+//          .flatMap {
+//            kv =>
+//              kv._2
+//                .map(new String(_, Utf8))
+//                .flatMap {
+//                  c => if (Charset.isSupported(c)) Some(Charset.forName(c)) else None
+//                }
+//          }.getOrElse(defaultEncoding)
+//
+//      kvPairs.map {
+//        case (key, value) => decode(new String(key, charset)) -> value.map(new String(_, charset)).map(decode)
+//      }
+//    }
+//  }
+  // commented out because I still don't understand how charset works here, in conjunction with assuming proper input
+  // to PercentEncodedString
+}
```

Comment:
I think we need this verison of `w3cHtml5FormUrlDecode`. Otherwise we can only decode forms that were submitted with the UTF-8 charset (really this is an issue with `unsafePercentDecode` but I suspect dropping its charset parameter was motivated here).